homeassistant/components/enphase_envoy/sensor.py

```diff
@@ -65,53 +76,94 @@ async def async_setup_platform(hass, config, async_add_entities, discovery_info=
 
     envoy_reader = EnvoyReader(ip_address, username, password)
 
+    async def async_update_data():
+        """Fetch data from API endpoint."""
+        async with async_timeout.timeout(30):
+            try:
+                data = await envoy_reader.update()
+            except httpcore.ProtocolError as err:
+                raise UpdateFailed(f"Error communicating with API: {err}") from err
+            except httpcore.ConnectTimeout as err:
+                raise UpdateFailed(f"Timeout error with API: {err}") from err
+
+            _LOGGER.debug("Retrieved data from API: %s", data)
+
+            if "can't handle event type ConnectionClosed" in str(
+                data.get("inverters_production")
+            ):
+                raise UpdateFailed("Inverter updated failed")
+
+            _LOGGER.debug("Returning API data.")
+            return data
+
+    coordinator = DataUpdateCoordinator(
+        homeassistant,
+        _LOGGER,
+        name="sensor",
+        update_method=async_update_data,
+        update_interval=SCAN_INTERVAL,
+    )
+
+    await coordinator.async_refresh()
+
     entities = []
     # Iterate through the list of sensors
     for condition in monitored_conditions:
-        if condition == "inverters":
-            try:
-                inverters = await envoy_reader.inverters_production()
-            except requests.exceptions.HTTPError:
-                _LOGGER.warning(
-                    "Authentication for Inverter data failed during setup: %s",
-                    ip_address,
-                )
-                continue
-
+        if condition == "inverters" and coordinator.data is not None:
+            inverters = coordinator.data.get("inverters_production")
+            _LOGGER.debug("Inverter data: %s", inverters)
             if isinstance(inverters, dict):
                 for inverter in inverters:
                     entities.append(
                         Envoy(
-                            envoy_reader,
                             condition,
                             f"{name}{SENSORS[condition][0]} {inverter}",
                             SENSORS[condition][1],
+                            coordinator,
                         )
                     )
-
-        else:
+                    _LOGGER.debug(
+                        "Adding inverter SN: %s - Type: %s.",
+                        f"{name}{SENSORS[condition][0]} {inverter}",
+                        condition,
+                    )
+            elif "Unable to connect to Envoy" in inverters:
+                _LOGGER.error(
+                    "Unable to connect to Enphase Envoy during setup. Inverter entities not added. Please check IP address and credentials are correct."
+                )
+            elif "can't handle event type ConnectionClosed" in inverters:
+                _LOGGER.error(
+                    "Communication error with Enphase Envoy during setup. Inverter entities not added."
+                )
```

Comment:
Can we use a typed exceptions here instead to avoid the string matching?

Reply:
The api being used is returning a string with the exception inside that. I know not the best solution but will need to be addressed. Without changing the library. I only added these because they were not being caught in the previous sensor code. The api is returning the same as it was previously. I can remove the string checks and go back to the previous behavior of the sensor code which was to really not check much?

Reply:
Changed to using exceptions rather than string matching.